daemon/api.go

```diff
@@ -2637,3 +2639,80 @@ func getLogs(c *Command, r *http.Request, user *auth.UserState) Response {
 		follow:     follow,
 	}
 }
+
+type appInstruction struct {
+	Action string   `json:"action"`
+	Names  []string `json:"names"`
+	client.StartOptions
+	client.StopOptions
+	client.RestartOptions
+}
+
+func postApps(c *Command, r *http.Request, user *auth.UserState) Response {
+	var inst appInstruction
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&inst); err != nil {
+		return BadRequest("cannot decode request body into service operation: %v", err)
+	}
+	if len(inst.Names) == 0 {
+		// on POST, don't allow empty to mean all
+		return BadRequest("cannot perform operation on services without a list of services to operate on")
+	}
+
+	st := c.d.overlord.State()
+	appInfos, rsp := appInfosFor(st, inst.Names, appInfoOptions{service: true})
+	if rsp != nil {
+		return rsp
+	}
+	if len(appInfos) == 0 {
+		// can't happen: appInfosFor with a non-empty list of services
+		// shouldn't ever return an empty appInfos with no error response
+		return InternalError("no services found")
+	}
+
+	argv := make([]string, 2, 3+len(appInfos))
+	argv[0] = "systemctl"
+	argv[1] = inst.Action
+
+	switch inst.Action {
+	case "start":
+		if inst.Enable {
+			argv[1] = "enable"
+			argv = append(argv, "--now")
+		}
+	case "stop":
+		if inst.Disable {
+			argv[1] = "disable"
+			argv = append(argv, "--now")
+		}
+	case "restart":
+		if inst.Reload {
+			argv[1] = "try-reload-or-restart"
+		}
+	default:
+		return BadRequest("unknown action %q", inst.Action)
+	}
+
+	snapNames := make(map[string]bool, len(appInfos))
+	names := make([]string, len(appInfos))
+	for i, svc := range appInfos {
+		argv = append(argv, svc.ServiceName())
+		snapName := svc.Snap.Name()
+		names[i] = snapName + "." + svc.Name
+		snapNames[snapName] = true
+	}
+
+	desc := fmt.Sprintf("%s of %v", inst.Action, names)
+
+	st.Lock()
+	defer st.Unlock()
+	if err := snapstate.CheckChangeConflictMulti(st, snapNames, nil); err != nil {
+		return Conflict(err.Error())
```

Comment:
Hmm... perhaps. Let's see...

> ### 6.5.8.  409 Conflict
> The 409 (Conflict) status code indicates that the request could not be completed due to a conflict with the current state of the target resource.  This code is used in situations where the user might be able to resolve the conflict and resubmit the request.  The server SHOULD generate a payload that includes enough information for a user to recognize the source of the conflict.
> 
> Conflicts are most likely to occur in response to a PUT request.  For example, if versioning were being used and the representation being PUT included changes to a resource that conflict with those made by an earlier (third-party) request, the origin server might use a 409 response to indicate that it can't complete the request.  In this case, the response representation would likely contain information useful for merging the differences based on the revision history.

To me this matches pretty closely with this usage: the request can't be completed because it clashes with the current state, so we let the user know what they can do (in this case, wait for the other change to finish). Do I have this wrong?